client/src/components/profile/Profile.js

```diff
@@ -1,26 +1,44 @@
 import React from "react";
 import { connect } from "react-redux";
 import { Link } from "react-router-dom";
+import { fetchProfile, createProfile } from "../../actions";
 
 class Profile extends React.Component {
+  componentDidUpdate() {
+    if (this.props.currentUserId != null && this.props.profile == null) {
+      this.props.fetchProfile(this.props.currentUserId);
+    }
+  }
+
   render() {
-    let to = "/";
-    if (this.props.isSignedIn) {
-      to = `/profile/${this.props.currentUserId}`;
+    if (this.props.isSignedIn && this.props.profile != null) {
+      return (
+        <Link
+          to={`/profile/${this.props.currentUserId}`}
+          className="item"
+          style={{ color: "white" }}
+        >
+          {this.props.profile.username}
+        </Link>
+      );
+    }
+    else if (this.props.isSignedIn && this.props.profile == null){
```

Comment:
I might be wrong, but I think this line is causing the app to create a new user entry whenever the user signs in or refreshes the page, since the profile is null within the first few moments of the app launching.
For example:
- Sign in; An entry is created in the DB
- Change your name; The name doesn't actually update in the header, but that seems like a separate issue.
- Check the database; Your username is updated correctly
- Refresh the page; Your username is reset, and the DB is also updated with the default information

If you change your name, comment out this line, then refresh the page, the issue doesn't occur.

To remedy this, I think we should start off by creating the profile within the GoogleAuth component, which I think makes more sense, since that's where the sign in flow is.
The second thing that needs to happen I think, is that instead of checking if profile == null, you need to query the DB directly to see if an entry already exists with the ID. If it doesn't exist, create a profile, otherwise, don't (and possibly fetch the profile to put it into the global state, though I'm not too sure if that's desired).

Reply:
I still think that I should leave the profile stuff to the Profile component. I don't want to perform redundant fetches or have functionality scattered across components. But I did change the second thing you mentioned. So the change that has been made is that when you fetch the profile, if it fails, then it will create a new profile.

Reply:
That is fair. But I think this approach might still have issues. What if there's just a random error when fetching the data? Will this create a duplicate profile?

I think what you want here is:
- Query the db for a user with the userid
- if (!querySnapshot.exists()) { create profile }
- else { just return the profile }

Reply:
Yeah I agree with Emilio. You wouldn't want to create a profile on just any error during the fetch.

Reply:
Just to add. I still think that creating the profile should probably be done within the signIn action. All the information you need is there, and it would make fetchProfile do exactly what it says on the tin, instead of having extra functionality that the name doesn't suggest.

Reply:
I think it should be in the signIn as well since if the create profile fails on a users initial login we need to be able to show an error and not log them in. Otherwise we will run into other errors if a user logs in and no profile was created.

Reply:
Ok. Changed again as per Emilio's suggestion. Now checks if querysnapshot is empty before creating a new profile.

Reply:
Ok, so I made a new status where it is waiting for sign-in. I think theres some confusion about which sign in is which. The way I see it theres 3 sign-ins, and all 3 needs to be fulfilled for the user to be truly logged in.
1. Google sign in
2. Local sign in (via the Redux store state: isSignedIn)
3. Firebase sign in (retrieving user info via the database)

So the new changes would sign you into `1`, but not into `2` until `3` has been resolved whether by creating a new profile or fetching a profile. This means that either failing to create a profile or failing to fetch a profile will log the user out of whatever they have whichever of `1`, `2`, and/or `3` they have already logged into.

Also, as Emilio suggested, it will only create a new user if querySnapshot is empty. Otherwise it returns the profile. 

However, I left the code for createProfile within a conditional in the fetchProfile (although I moved the code into its own function for less clutter) because otherwise it will slow down the whole process and probably make more clutter due to the asynchronous nature of the calls. 
(i.e: fetchProfile would need to dispatch an action to the store to change a flag saying there is no profile, wait for the flag to update, wait for the signIn to pickup on the update, and then dispatch another action for createProfile separately from the signIn)
I have actually tried this way and it looks confusing as hell.

Also, I fixed the issue of the username not updating. It updates instantly when you edit your profile now.